In [4]:
import numpy as np

big_integer = int(1e6)
d_k = 25

In [5]:
mean = 0
variance = 1

q = np.random.normal(mean, variance, (big_integer, d_k))
print(f"{np.mean(q):.2f}    {np.std(q):.2f}")

k = np.random.normal(mean, variance, (big_integer, d_k))
print(f"{np.mean(k):.2f}    {np.std(k):.2f}")

-0.00    1.00
-0.00    1.00


In [6]:
qk = q * k             # qk.shape = (big_integer, d_k)
qk = (qk).sum(axis=1)  # qk.shape = (big_integer, )
print(f"{np.mean(qk):.4f}    {np.std(qk):.4f}")

0.0036    4.9991


4.9991 is almost equal to $\sqrt{d_k} = \sqrt{25} = 5$

> To illustrate why the dot products get large, assume that the components of $q$ and $k$ are independent random variables with mean 0 and variance 1. Then their dot product, $q \cdot k = \sum_{i=1}^{d_k} q_ik_i$, has mean 0 and variance $d_k$.

So the statement from the paper "Attention is All You Need" is slightly erroneous. The variance should br $\sqrt{d_k}$ instead of $d_k$.